plugins/hls-selection-range-plugin/src/Ide/Plugin/SelectionRange/ASTPreProcess.hs

```diff
@@ -0,0 +1,156 @@
+import qualified Data.Set                        as Set
+import           Development.IDE.GHC.Compat      (ContextInfo (MatchBind, TyDecl, ValBind),
+                                                  HieAST (..), Identifier,
+                                                  IdentifierDetails (identInfo),
+                                                  NodeInfo (NodeInfo, nodeIdentifiers),
+                                                  RealSrcSpan, RefMap, Span,
+                                                  combineRealSrcSpans,
+                                                  flattenAst,
+                                                  isAnnotationInNodeInfo,
+                                                  mkAstNode, nodeInfoFromSource,
+                                                  realSrcSpanEnd,
+                                                  realSrcSpanStart)
+import           Development.IDE.GHC.Compat.Util (FastString)
+import           Prelude                         hiding (span)
+
+{-|
+Extra arguments for 'preaProcessAST', meant to be used in a 'Reader' context. We use 'Reader' to combine
+-}
+newtype PreProcessEnv a = PreProcessEnv
+    { preProcessEnvRefMap :: RefMap a
+    }
+
+{-|
+Before converting the HieAST to selection range, we need to run some passes on it. Each pass potentially modifies
+the AST to handle some special cases.
+
+'preProcessAST' combines the passes. Refer to 'mergeImports' or 'mergeSignatureWithDefinition' as
+a concrete example example.
+
+Adding another manipulation to the AST is simple, just implement a function of type
+`HieAST a -> Reader (PreProcessEnv a) (HieAST a)`, and append it to 'preProcessAST' with `>>=`.
+
+If it goes more complex, it may be more appropriate to split different manipulations to different modules.
+-}
+preProcessAST :: HieAST a -> Reader (PreProcessEnv a) (HieAST a)
+preProcessAST node = mergeImports node >>= mergeSignatureWithDefinition
+
+{-|
+Combines adjacent import declarations under a new parent node, so that the user will have an extra step selecting
+the whole import area while expanding/shrinking the selection range.
+-}
+mergeImports :: forall a. HieAST a -> Reader (PreProcessEnv a) (HieAST a)
+mergeImports node = pure $ node { nodeChildren = children }
+  where
+    children :: [HieAST a]
+    children = mapMaybe merge
+        . groupBy (\x y -> nodeIsImport x && nodeIsImport y)
+        . nodeChildren $ node
+
+    merge :: [HieAST a] -> Maybe (HieAST a)
+    merge []  = Nothing
+    merge [x] = Just x
+    merge xs  = createVirtualNode xs
+
+nodeIsImport :: HieAST a -> Bool
+nodeIsImport = isAnnotationInAstNode ("ImportDecl", "ImportDecl")
+
+createVirtualNode :: [HieAST a] -> Maybe (HieAST a)
+createVirtualNode [] = Nothing
+createVirtualNode children = Just $ mkAstNode (NodeInfo mempty mempty mempty) span' children
+  where
+    span' :: RealSrcSpan
+    span' = foldl1' combineRealSrcSpans . fmap nodeSpan $ children
```

Comment:
Nit, but we're trying to avoid partial functions in HLS. You could make `createVirtualNode` take a `NonEmpty,` then it won't need to return `Maybe` and you can use a safe `fold1`.

Reply:
Nice advice!